src/components/views/settings/tabs/room/PeopleRoomSettingsTab.tsx

```diff
@@ -0,0 +1,175 @@
+    };
+
+    const onError = (): void => setDisabled(false);
+
+    return (
+        <div className="mx_PeopleRoomSettingsTab_knock">
+            <MemberAvatar height={42} member={roomMember} width={42} />
+            <div className="mx_PeopleRoomSettingsTab_content">
+                <span className="mx_PeopleRoomSettingsTab_name">{roomMember.name}</span>
+                <Timestamp roomMember={roomMember} />
+                <span className="mx_PeopleRoomSettingsTab_userId">{roomMember.userId}</span>
+                <SeeMoreOrLess roomMember={roomMember} />
+            </div>
+            <AccessibleButton
+                className="mx_PeopleRoomSettingsTab_action"
+                disabled={!canKick || disabled}
+                kind="icon_primary_outline"
+                onClick={() => handleDeny(roomMember.userId)}
+                title={_t("Deny")}
+            >
+                <XIcon width={18} height={18} />
+            </AccessibleButton>
+            <AccessibleButton
+                className="mx_PeopleRoomSettingsTab_action"
+                disabled={!canInvite || disabled}
+                kind="icon_primary"
+                onClick={() => handleApprove(roomMember.userId)}
+                title={_t("Approve")}
+            >
+                <CheckIcon width={18} height={18} />
+            </AccessibleButton>
+        </div>
+    );
+};
+
+export const PeopleRoomSettingsTab: VFC<{ room: Room }> = ({ room }) => {
+    const [knockMembers, setKnockMembers] = useState(room.getMembersWithMembership("knock"));
+    const client = room.client;
+    const userId = client.getUserId() || "";
+    const canInvite = room.canInvite(userId);
+    const member = room.getMember(userId);
+    const state = room.getLiveTimeline().getState(EventTimeline.FORWARDS);
+    const canKick = member && state ? state.hasSufficientPowerLevelFor("kick", member.powerLevel) : false;
+    const roomId = room.roomId;
+
+    const handleApprove = (userId: string): Promise<void> =>
+        new Promise((_, reject) =>
+            client.invite(roomId, userId).catch((error) => {
+                onError(error);
+                reject();
+            }),
+        );
+
+    const handleDeny = (userId: string): Promise<void> =>
+        new Promise((_, reject) =>
+            client.kick(roomId, userId).catch((error) => {
+                onError(error);
+                reject();
+            }),
+        );
+
+    const onError = (error: MatrixError): IHandle<typeof ErrorDialog> =>
+        Modal.createDialog(ErrorDialog, {
+            title: error.name,
+            description: error.message,
+        });
+
+    useEffect(() => {
+        const updateKnockMembers = (): void => setKnockMembers(room.getMembersWithMembership("knock"));
+        client.on(RoomStateEvent.Members, updateKnockMembers);
+        return () => {
+            client.removeListener(RoomStateEvent.Members, updateKnockMembers);
+        };
+    }, [client, room]);
```

Comment:
Should we use `useTypedEventEmitterState` here instead?

Reply:
Thanks, for pointing that out 🙂